tutorial/spiders/pakistan_stores_spider.py

```diff
@@ -0,0 +1,97 @@
+"""
+This module contains a spider of pakistanistores.com site
+that crawls and extracts data
+"""
+import json
+from functools import partial
+
+import scrapy
+
+
+class PakistanStoresSpider(scrapy.Spider):
+    """
+    This class crawls the site pakistanistores.com and extracts data
+    """
+
+    name = "pakistan_stores"
+    current_page = 1
+    products = []
+    processed_products = 0
+
+    def start_requests(self):
+        url = getattr(self, "url", None)
+        yield scrapy.Request(url, self.parse)
+
+    def parse(self, response, **kwargs):
+        """
+        This function crawls through each page and extracts all product information
+        Args:
+            response(TextReponse): Contains the page contents
+        Returns:
+            (Any or None):  A request to another page
+                            Or
+                            None if everything processed
+        """
+        last_page = response.css("a.page-link.navigate::attr(data-href)")[-1].get()
+        last_page_number = int(last_page[last_page.find("=") + 1 :])
+        product_containers = response.css("li.col-md-3.col-md-3.col-sm-6.col-xs-6 a")
+        for product_container in product_containers:
+            name = product_container.attrib["title"]
+            url = response.urljoin(product_container.attrib["href"])
```

Comment:
Let's name it `page_url` for better readability.

Reply:
Done